### ISSUE TITLE
[12.0] ON CONFLICT DO UPDATE command cannot affect row a second time

### ISSUE THREAD

**Issue by marcelsavegnago** [odoo/odoo#41364]
Impacted versions: 12 

Steps to reproduce: Load pt_BR language.
(I simulated the problem within an instance of Odoo in Runbot)
```
Odoo Server Error
Traceback (most recent call last):
  File "/data/build/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/data/build/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/data/build/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/data/build/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/data/build/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/data/build/odoo/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/data/build/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/data/build/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/data/build/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/data/build/odoo/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/data/build/odoo/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/data/build/odoo/addons/website/wizard/base_language_install.py", line 25, in lang_install
    action = super(BaseLanguageInstall, self).lang_install()
  File "/data/build/odoo/odoo/addons/base/wizard/base_language_install.py", line 36, in lang_install
    mods.with_context(overwrite=self.overwrite)._update_translations(self.lang)
  File "/data/build/odoo/odoo/addons/base/models/ir_module.py", line 885, in _update_translations
    self.env['ir.translation'].load_module_terms(mod_names, filter_lang)
  File "/data/build/odoo/odoo/addons/base/models/ir_translation.py", line 825, in load_module_terms
    tools.trans_load(self._cr, trans_file, lang, verbose=False, module_name=module_name, context=context)
  File "/data/build/odoo/odoo/tools/translate.py", line 1016, in trans_load
    result = trans_load_data(cr, fileobj, fileformat, lang, verbose=verbose, module_name=module_name, context=context)
  File "/data/build/odoo/odoo/tools/translate.py", line 1160, in trans_load_data
    irt_cursor.finish()
  File "/data/build/odoo/odoo/addons/base/models/ir_translation.py", line 156, in finish
    """ % (self._model_table, self._table))
  File "/data/build/odoo/odoo/sql_db.py", line 148, in wrapper
    return f(self, *args, **kwargs)
  File "/data/build/odoo/odoo/sql_db.py", line 225, in execute
    res = self._obj.execute(query, params)
psycopg2.ProgrammingError: ON CONFLICT DO UPDATE command cannot affect row a second time
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
```

**Comment by Yenthe666** [comment]
Pretty much a duplicate of https://github.com/odoo/odoo/issues/32801 and since somebody mentioned the translation issue there too I'm closing this one in favor of https://github.com/odoo/odoo/issues/32801

**Comment by mart-e** [comment]
No not a duplicate. This should never happen when loading translations. It means a .pot or .po file is incorrect and must be fixed.

**Comment by yucer** [comment]
@mart-e Do you think that an error loading a different translation for a term it is so important to stop one update. I mean it should be one STOP error ?

It sound like a warning to me. If the translation of a `.pot` is different to the one in `.po` I would make one of them to prevail and report it as warning.

I think it depends on the quality level that needs to be ensured. Some implementations might not consider a translation of a term so important to make the server fail. It might be one option.

In this [patch](https://gist.github.com/yucer/0c68310f8ce282fbbd3f61db545478c9) I do report the error as Warning and later delete the duplicated / conflicting translations to be imported.

But it was just to make my point. I am sure you have a better idea of  how to handle that.

**Comment by mart-e** [comment]
@yucer this is debatable but yes, if there is an error, I think we should see it and fix it as soon as possible. I do not think it is useful to log the problems, this is not a incentive strong enough to fix it.

We are working at #41639 and #41650 to fix both cases without removing the duplicates.

**Comment by yucer** [comment]
Just to add more info for this specific context, that might be useful.  As long as I see:

> The problem occur once the code of the import wizard has collected all those temporary rows to update and there is more than one translation for the same term in them.

The specific cases that I have seen is one coming from the `.po` and one coming from the `.pot`. That means that either the `.po` or the `.pot` are not updated.  Unfortunately I didn't realize how the files came to this conflict, the cases seem to come from **Transiflex** commits.

**Comment by yucer** [comment]
> @yucer this is debatable but yes, if there is an error, I think we should see it and fix it as soon as possible. I do not think it is useful to log the problems, this is not a incentive strong enough to fix it.
> 

But, then the test jobs should load all the possible languages. I we you don't want a commit to be tested and reported as *good* to be merged with these translation errors that would trigger update errors later when loading the translation for a language.

Some of these errors did happen in `base` module, and in **French** language. I guess that it is a very common combination used even in your *SaaS* instances.

So... if the translation error in a third language is so important fail... and the translation data migth come from the code, then the test jobs should test the third languages also. I guess.

**Comment by mart-e** [comment]
> I didn't realize how the files came to this conflict, the cases seem to come from Transiflex commits

How so? Transifex should not introduce issues. The errors are usually introduced by bad .pot file that was not correctly updated (manually).

> the test jobs should load the others languages also

Could be a good idea, feel free to submit new test as an improvement and I will review it.

**Comment by yucer** [comment]
> > I didn't realize how the files came to this conflict, the cases seem to come from Transiflex commits
> 
> How so? Transifex should not introduce issues. The errors are usually introduced by bad .pot file that was not correctly updated (manually).
> 

Check the example referenced [here](https://github.com/odoo/odoo/issues/32801#issuecomment-561169447). 

The commit that introduced the error was [[I18N] Update translation terms from Transifex](https://github.com/odoo/odoo/commit/5db13a3a61f722447365b28feab1daff6cbd2578) . Was it a manual change ?

The conflict was with [stock](https://github.com/odoo/odoo/blob/5db13a3a61f722447365b28feab1daff6cbd2578/odoo/addons/base/i18n/fr.po#L28461) in *fr.po* and [Stock](https://github.com/odoo/odoo/blob/5db13a3a61f722447365b28feab1daff6cbd2578/odoo/addons/base/i18n/base.pot#L19286) in *base.pot*

**Comment by mart-e** [comment]
No actually the issue was present since ec5980f658eec09872294d32e when 2 terms had the same source in `base.pot`. The `.po` files will use the `.pot` as reference, so translations not present in the reference will be ignored when parsing the po file.
Closing as fixed at  #41650
